runtime/src/comm/ofi/comm-ofi.c

```diff
@@ -2915,21 +3114,136 @@ void ofi_put_ll(const void* addr, c_nodeid_t node,
                                mrDesc, rxRmaAddr(tcip, node),
                                mrRaddr, mrKey, ctx),
                       checkTxCQ(tcip));
-  if (tcip->txCntr != NULL) {
-    tcip->numTxnsBegun++;
+  tcip->numTxns++;
+  tciFree(tcip);
+}
+
+
+static
+void ofi_put_V(int v_len, void** addr_v, void** local_mr_v,
+               c_nodeid_t* locale_v, void** raddr_v, uint64_t* remote_mr_v,
+               size_t* size_v) {
+  DBG_PRINTF(DBG_RMA | DBG_RMAWRITE,
+             "PUT_V(%d) %d:%p <= %p, size %zd",
+             v_len, (int) locale_v[0], raddr_v[0], addr_v[0], size_v[0]);
+
+  struct perTxCtxInfo_t* tcip;
+  CHK_TRUE((tcip = tciAlloc()) != NULL);
+
+  if (tcip->numTxns >= txCQLen) {
+    while (tcip->numTxns > 0) {
+      checkTxCQ(tcip);
+    }
   }
+
+  for (int vi = 0; vi < v_len; vi++) {
+    struct iovec iov = (struct iovec)
+                       { .iov_base = addr_v[vi],
+                         .iov_len = size_v[vi] };
+    struct fi_rma_iov riov = (struct fi_rma_iov)
+                             { .addr = (uint64_t) raddr_v[vi],
+                               .len = size_v[vi],
+                               .key = remote_mr_v[vi] };
+    struct fi_msg_rma msg = (struct fi_msg_rma)
+                            { .msg_iov = &iov,
+                              .desc = &local_mr_v[vi],
+                              .iov_count = 1,
+                              .addr = rxRmaAddr(tcip, locale_v[vi]),
+                              .rma_iov = &riov,
+                              .rma_iov_count = 1,
+                              .context = NULL,
+                              .data = 0 };
+    DBG_PRINTF(DBG_RMA | DBG_RMAWRITE,
+               "tx writemsg: %d:%p <= %p, size %zd, key 0x%" PRIx64,
+               (int) locale_v[vi], (void*) riov.addr, iov.iov_base,
+               iov.iov_len, riov.key);
+    if (++(tcip->numTxns) >= txCQLen || vi == v_len - 1) {
+      // Initiate last transaction in group and wait for whole group.
+      OFI_RIDE_OUT_EAGAIN(fi_writemsg(tcip->txCtx, &msg, 0),
+                          checkTxCQ(tcip));
+      while (tcip->numTxns > 0) {
+        checkTxCQ(tcip);
+      }
+    } else {
+      // Add another transaction to the group and go on without waiting.
+      OFI_RIDE_OUT_EAGAIN(fi_writemsg(tcip->txCtx, &msg, FI_MORE),
+                          checkTxCQ(tcip));
+    }
+  }
+
   tciFree(tcip);
 }
 
 
+/*
+ *** START OF BUFFERED PUT OPERATIONS ***
+ *
+ * Support for buffered PUT operations. We internally buffer PUT operations and
+ * then initiate them with chained transactions for increased transaction rate.
```

Comment:
Not using chained transactions.

Reply:
I chose to maximize similarity with the comm=ugni code in order to make it easier to find shared code in the (hopefully near) future when we move much of this into a common place.  But I wouldn't be averse to changing this so it's accurate.

Reply:
Keeping it as-is works for me